lib/streamlit/runtime/session_manager.py

```diff
@@ -0,0 +1,338 @@
+    session: AppSession
+    script_run_count: int = 0
+
+
+class SessionStorageError(Exception):
+    """Exception class for errors raised by SessionStorage.
+
+    The original error that causes a SessionStorageError to be (re)raised will generally
+    be an I/O error specific to the concrete SessionStorage implementation.
+    """
+
+
+class SessionStorage(Protocol):
+    @abstractmethod
+    def get(self, session_id: str) -> Optional[SessionInfo]:
+        """Return the SessionInfo corresponding to session_id, or None if one does not
+        exist.
+
+        Parameters
+        ----------
+        session_id
+            The unique ID of the session being fetched.
+
+        Returns
+        -------
+        SessionInfo or None
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to fetch the session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def save(self, session_info: SessionInfo) -> None:
+        """Save the given session.
+
+        Parameters
+        ----------
+        session_info
+            The SessionInfo being saved.
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while saving the given session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def delete(self, session_id: str) -> None:
+        """Mark the session corresponding to session_id for deletion and stop tracking
+        it.
+
+        Note that:
+          * Calling delete on an ID corresponding to a nonexistent session is a no-op.
+          * Calling delete on an ID should cause the given session to no longer be
+            tracked by this SessionStorage, but exactly when and how the session's data
+            is eventually cleaned up is a detail left up to the implementation.
+
+        Parameters
+        ----------
+        session_id
+            The unique ID of the session to delete.
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to delete the session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def list(self) -> List[SessionInfo]:
+        """List all sessions tracked by this SessionStorage.
+
+        Returns
+        -------
+        List[SessionInfo]
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to list sessions.
+        """
+        raise NotImplementedError
+
+
+class SessionManager(Protocol):
+    """SessionManagers are responsible for encapsulating all session lifecycle behavior
+    that the Streamlit Runtime may care about.
+
+    A SessionManager must define the following required methods:
+      * __init__
+      * connect_session
+      * close_session
+      * get_session_info
+      * list_sessions
+
+    SessionManager implementations may also choose to define the notions of active and
```

Comment:
It's useful to note that the motivation for supporting active/inactive sessions is the current state of the world where sessions are tightly coupled to websocket connections. Those with an open connection are active, and those without are inactive.

In the eventually future, we may have a notion of session that isn't so tightly coupled to websockets, at which point differentiating between the two may no longer be useful -- any existing session would be an active one.

Reply:
Wow -- who is this person and how does he know so much about what my thought process is

![spiderman meme](https://i.kym-cdn.com/entries/icons/mobile/000/023/397/C-658VsXoAo3ovC.jpg)